### DISCUSSION TITLE
Python Api / Plugins / Auto song search

### DISCUSSION THREAD

**Discussion by Craig-Venables** [nicotine-plus/nicotine-plus#2397] [Q&A]
Hi

I'm just wondering if there is an API for nicotine-plus? 
Is there any way I could use python to automatically search for songs etc... I have had a look around but can't seem to find anything on the topic, I did find the official python page (https://pypi.org/project/nicotine-plus/) but there is no documentation? I didn't know whether id need to create my own plugin or ifs just not possible at all?

I'm very new to using python and coding in general so I apologize if this is a very stupid question for context I'm trying to search for a specific list of songs and download the top result if it reaches specific criteria.

Cheers for any help, Craig

**Comment by mathiascode** [comment]
The API is essentially everything in the core (https://github.com/nicotine-plus/nicotine-plus/tree/master/pynicotine). We've been working on making more things available to use for plugins, but there will most likely never be a fully stable/frozen and documented API.

Note that https://pypi.org/project/nicotine-plus/ is the full Nicotine+ client, not a library. Nicotine+ is just available on PyPi for convenience, so people can install it with `pip install nicotine-plus`.

I believe the server might ban you if you perform too many searches at once, but I think you could accomplish what you want with a plugin:

- Perform a search: https://github.com/nicotine-plus/nicotine-plus/blob/2a90903fb96e5d234cf97128907456d8ea106581/pynicotine/search.py#L189
- Listen to the "file-search-response" event, similar to this: https://github.com/nicotine-plus/nicotine-plus/blob/2a90903fb96e5d234cf97128907456d8ea106581/pynicotine/search.py#L52-L60 https://github.com/nicotine-plus/nicotine-plus/blob/2a90903fb96e5d234cf97128907456d8ea106581/pynicotine/search.py#L325
- Check specific criteria in the response message (msg): https://github.com/nicotine-plus/nicotine-plus/blob/2a90903fb96e5d234cf97128907456d8ea106581/pynicotine/slskmessages.py#L2693
- Download a specific file with get_file(): https://github.com/nicotine-plus/nicotine-plus/blob/2a90903fb96e5d234cf97128907456d8ea106581/pynicotine/transfers.py#L1589

Note that this depends on the new event system, which in only available in master, and has not been optimized for plugins yet: https://github.com/nicotine-plus/nicotine-plus/discussions/1540#discussioncomment-4720241

To get you started, I could try to create a basic plugin for you based on these steps.

**Reply by Craig-Venables** [reply]
Excellent thank you, I'll have a look into what you mentioned above and see if I can get anything to work. ill probably delay each search by a second to hopefully prevent a ban
